### PR TITLE
trying to do version

### PR DESCRIPTION
**Pull request by gaurimadhok**
This PR is trying to give users the ability to decide what Docker version to use in the mixin.

### CODE REVIEW

File: pkg/docker/build.go [hunk 1]
```diff
@@ -12,6 +15,7 @@ ARG BUNDLE_DIR
 RUN apt-get update && apt-get install -y curl ca-certificates
 
 ARG DOCKER_VERSION=19.03.8
+RUN curl -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/docker-${DOCKER_VERSION}.tgz && \
```
Comment (vdice) [inline pkg/docker/build.go:18]: This added line appears to duplicate the one below, right?

Comment (vdice) [reply pkg/docker/build.go:18]: On line 17 above, I think `ARG DOCKER_VERSION=19.03.8` should be updated to `ARG DOCKER_VERSION=%s` now that we intend to use the string as a template with *printf.

File: pkg/docker/build.go [hunk 2]
```diff
@@ -21,9 +25,37 @@ RUN curl -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/do
...
+	})
+	if err != nil {
+		return err
+	}
+
+	if input.Config.ClientVersion != "" {
+		m.DockerVersion = input.Config.ClientVersion
+	}
+
+	log.Println(m.Out, dockerfileLines, m.DockerVersion, m.WorkingDir)
```
Comment (vdice) [inline pkg/docker/build.go:57]: I believe we want this to print to `m.Out` and it should be updated to use `printf` in tandem with the preceding note (adding a `%s` formatter).  So, something like:

`fmt.Fprintf(m.Out, dockerfileLines, m.DockerVersion)`

(Note, I also dropped `m.WorkingDir` as it doesn't currently apply for this mixin.)

File: pkg/docker/build.go (hunk 1 above)
Comment (gaurimadhok) [reply pkg/docker/build.go]: thank you, that makes sense!

File: pkg/docker/build.go (hunk 2 above)
Comment (gaurimadhok) [reply pkg/docker/build.go]: got it!

**Comment by gaurimadhok** [comment]
After making the changes you suggested @vdice, the custom docker version test passed but the default one did not. I added lines 54-56 in build.go in order to get the default version to be set - does that look right? I wasn't sure if that is the way you are supposed to make sure it is set.

File: pkg/docker/build.go [hunk 3]
```diff
@@ -21,9 +24,39 @@ RUN curl -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/do
...
+
+	if input.Config.ClientVersion != "" {
+		m.DockerVersion = input.Config.ClientVersion
+	} else {
+		m.DockerVersion = defaultDockerVersion
+	}
+
+	fmt.Fprintf(m.Out, dockerfileLines, m.DockerVersion)
 	return nil
+	//fmt.Fprintf(m.Out, dockerfileLines)
```
Comment (vdice) [inline pkg/docker/build.go:60]: Let's remove these two commented-out lines.

File: pkg/docker/build.go [hunk 4]
```diff
@@ -21,9 +24,39 @@ RUN curl -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/do
...
+		err := yaml.Unmarshal(contents, &input)
+		return &input, err
+	})
+	if err != nil {
+		return err
+	}
+
+	if input.Config.ClientVersion != "" {
+		m.DockerVersion = input.Config.ClientVersion
+	} else {
```
Comment (vdice) [inline pkg/docker/build.go:54]: Ah, I see the issue here.  Although this `else` is certainly fine, we can avoid it.

The reason we weren't seeing the default version unit test pass was b/c the `NewTestMixin()` helper didn't actually call `mixin.New()` which performs the action of setting the default.  In the process of adjusting that, I also updated the `mixin.New()` method signature to remove returning an `error`, to simplify logic.

Check it out: https://github.com/gaurimadhok/porter-docker/compare/set-version...vdice:gaurimadhok-set-version-plus-default-fix?expand=1

Feel free to use some/all of those changes.

Comment (gaurimadhok) [reply pkg/docker/build.go]: thank you that makes sense!

**Comment by gaurimadhok** [comment]
Currently I have the default version as 19.03.8, but the latest release is 19.03.12, so should I change the default to 19.03.12?

**Comment by vdice** [comment]
> Currently I have the default version as 19.03.8, but the latest release is 19.03.12, so should I change the default to 19.03.12?

No preference on my end.  I'll leave it up to you!

**Pull request merged**